OpenRA.Game/Graphics/TerrainSpriteLayer.cs

```diff
@@ -45,25 +65,58 @@ public TerrainSpriteLayer(World world, WorldRenderer wr, Sheet sheet, BlendMode
 			map = world.Map;
 			rowStride = 6 * map.MapSize.X;
 
-			vertices = new Vertex[rowStride * map.MapSize.Y];
-			vertexBuffer = Game.Renderer.Device.CreateVertexBuffer(vertices.Length);
+			// We can use a buffer to track updates. We can then only send updates in the visible region to the GPU.
+			// This saves CPU by only sending updated rows when they become visible. This requires keeping a full copy
+			// of the vertices, so it doubles the memory required.
+			this.buffered = buffered;
+
+			if (buffered)
+				dirtyRows = new HashSet<int>();
+
+			vertices = new Vertex[buffered ? rowStride * map.MapSize.Y : 6];
+			vertexBuffer = Game.Renderer.Device.CreateVertexBuffer(rowStride * map.MapSize.Y);
 			emptySprite = new Sprite(sheet, Rectangle.Empty, TextureChannel.Alpha);
 
 			wr.PaletteInvalidated += UpdatePaletteIndices;
 		}
 
 		void UpdatePaletteIndices()
 		{
-			// Everything in the layer uses the same palette,
-			// so we can fix the indices in one pass
-			for (var i = 0; i < vertices.Length; i++)
+			if (buffered)
 			{
-				var v = vertices[i];
-				vertices[i] = new Vertex(v.X, v.Y, v.Z, v.S, v.T, v.U, v.V, palette.TextureIndex, v.C);
+				// Everything in the layer uses the same palette,
+				// so we can fix the indices in one pass
+				for (var i = 0; i < vertices.Length; i++)
+				{
+					var v = vertices[i];
+					vertices[i] = new Vertex(v.X, v.Y, v.Z, v.S, v.T, v.U, v.V, palette.TextureIndex, v.C);
+				}
+
+				for (var row = 0; row < map.MapSize.Y; row++)
+					dirtyRows.Add(row);
 			}
+			else
+			{
+				var vertexBufferLength = rowStride * map.MapSize.Y;
+				var buffer = new Vertex[Math.Min(2048, vertexBufferLength)];
+
+				var start = 0;
+				while (start < vertexBufferLength)
+				{
+					var length = Math.Min(buffer.Length, vertexBufferLength - start);
+					vertexBuffer.GetData(buffer, start, length);
```

Comment:
Consider separating the palette indices into another buffer so you don't have to pull back vertices. This is potentially very bad, esp done piecemeal.

Need perf impact assessed on a variety of drivers.

Reply:
Let me know if you'd still like to know the perf impact, given this affects only the map editor.